File: src/main/java/me/zeroeightsix/kami/command/commands/NBTCommand.java
package me.zeroeightsix.kami.command.commands;

import me.zeroeightsix.kami.command.Command;
import me.zeroeightsix.kami.command.syntax.ChunkBuilder;
import me.zeroeightsix.kami.command.syntax.parsers.EnumParser;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;

import java.awt.*;
import java.awt.datatransfer.Clipboard;
import java.awt.datatransfer.StringSelection;

import static me.zeroeightsix.kami.util.MessageSendHelper.sendChatMessage;
import static me.zeroeightsix.kami.util.MessageSendHelper.sendErrorMessage;

/**
 * @author d1gress/Qther
 * Created by d1gress/Qther on 4/12/2019.
 */
public class NBTCommand extends Command {

    public NBTCommand() {
        super("nbt", new ChunkBuilder()
                .append("action", true, new EnumParser(new String[]{"get", "copy", "wipe"}))
                .build());
        setDescription("Does NBT related stuff (&fget&7, &fcopy&7, &fset&7)");
    }

    private final Clipboard clipboard = Toolkit.getDefaultToolkit().getSystemClipboard();
    private StringSelection nbt;

    @Override
    public void call(String[] args) {
        if (args[0].isEmpty()) {
            sendErrorMessage("Invalid Syntax!");
            return;
        }
        ItemStack item = mc.player.inventory.getCurrentItem();

        if (args[0].equalsIgnoreCase("get")) {
            if (item.getTagCompound() != null) {
                sendChatMessage("&6&lNBT:\n" + item.getTagCompound() + "");
            } else sendErrorMessage("No NBT on " + item.getDisplayName());
        } else if (args[0].equalsIgnoreCase("copy")) {
            if (item.getTagCompound() != null) {
                nbt = new StringSelection(item.getTagCompound() + "");
                clipboard.setContents(nbt, nbt);
                sendChatMessage("&6Copied\n&f" + (item.getTagCompound() + "\n") + "&6to clipboard.");
            } else sendErrorMessage("No NBT on " + item.getDisplayName());
        } else if (args[0].equalsIgnoreCase("wipe")) {
            sendChatMessage("&6Wiped\n&f" + (item.getTagCompound() + "\n") + "&6from " + item.getDisplayName() + ".");
            item.setTagCompound(new NBTTagCompound());
        }
    }
}

File: src/main/java/me/zeroeightsix/kami/command/commands/NBTCommand.kt
package me.zeroeightsix.kami.command.commands

import me.zeroeightsix.kami.command.ClientCommand
import me.zeroeightsix.kami.event.SafeExecuteEvent
import me.zeroeightsix.kami.util.text.MessageSendHelper
import me.zeroeightsix.kami.util.text.formatValue
import net.minecraft.item.ItemStack
import net.minecraft.nbt.NBTTagCompound

object NBTCommand : ClientCommand(
    name = "nbt",
    description = "Get, copy, paste, clear NBT for item held in main hand"
) {

    private var copiedNbtTag: NBTTagCompound? = null

    init {
        literal("get") {
            executeSafe {
                val itemStack = getHelpItemStack() ?: return@executeSafe
                val nbtTag = getNbtTag(itemStack) ?: return@executeSafe

                MessageSendHelper.sendChatMessage("NBT tags on item ${formatValue(itemStack.displayName)}")
                MessageSendHelper.sendRawChatMessage(nbtTag.toString())
            }
        }

        literal("copy") {
            executeSafe {
                val itemStack = getHelpItemStack() ?: return@executeSafe
                val nbtTag = getNbtTag(itemStack) ?: return@executeSafe

                copiedNbtTag = nbtTag
                MessageSendHelper.sendChatMessage("Copied NBT tags from item ${formatValue(itemStack.displayName)}")
            }
        }

        literal("paste") {
            executeSafe {
                val itemStack = getHelpItemStack() ?: return@executeSafe
                val nbtTag = copiedNbtTag ?: run {
                    MessageSendHelper.sendChatMessage("No copied NBT tags!")
                    return@executeSafe
                }

                itemStack.tagCompound = nbtTag
                MessageSendHelper.sendChatMessage("Pasted NBT tags to item ${formatValue(itemStack.displayName)}")
            }
        }

        literal("clear", "wipe") {
            executeSafe {
                val itemStack = getHelpItemStack() ?: return@executeSafe
                getNbtTag(itemStack) ?: return@executeSafe // Make sure it has a NBT tag before

                itemStack.tagCompound = NBTTagCompound()
            }
        }
    }

    private fun getNbtTag(itemStack: ItemStack): NBTTagCompound? {
        val nbtTag = itemStack.tagCompound

        if (nbtTag == null) {
            MessageSendHelper.sendChatMessage("Item ${formatValue(itemStack.displayName)} doesn't have NBT tag!")
        }

        return nbtTag
    }

    private fun SafeExecuteEvent.getHelpItemStack(): ItemStack? {
        val itemStack = player.inventory?.getCurrentItem()

        if (itemStack == null || itemStack.isEmpty) {
            MessageSendHelper.sendChatMessage("Not holding an item!")
            return null
        }

        return itemStack
    }

}